src/byte_sink.rs

```diff
@@ -0,0 +1,76 @@
+use std::io;
+
+use futures::{Async, Poll, Sink, StartSend, AsyncSink};
+
+use {AsyncWrite, AsyncRead};
+
+const BUFFER_SIZE: usize = 8 * 1024;
+
+/// A sink that consumes AsyncRead values, copying their contents to an
+/// AsyncWrite.
+pub struct ByteSink<W, R> {
+    write: W,
+    read: Option<R>,
+    write_cursor: usize,
+    read_cursor: usize,
+    buffer: [u8; BUFFER_SIZE],
+}
+
+/// Construct a sink that consumes AsyncReads, writing their entire contents to
+/// the supplied AsyncWrite.
+pub fn byte_sink<W, R>(write: W) -> ByteSink<W, R> {
+    ByteSink {
+        write: write,
+        read: None,
+        write_cursor: 0,
+        read_cursor: 0,
+        buffer: [0; BUFFER_SIZE],
+    }
+}
+
+impl<W: AsyncWrite, R: AsyncRead> Sink for ByteSink<W, R> {
+    type SinkItem = R;
+    type SinkError = io::Error;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if self.read.is_none() {
+            self.read = Some(item);
+            Ok(AsyncSink::Ready)
+        } else {
+            Ok(AsyncSink::NotReady(item))
```

Comment:
Currently the contract of `start_send` is that if it returns `NotReady` it "otherwise attempted `poll_complete` and still wasn't ready", could that be reflected here too?